scripts/resolve_deps/main.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/solo-io/ext-auth-plugin-examples/pkg/checks"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+)
+
+const (
+	errorReportFile     = "mismatched_dependencies.json"
+	suggestionsFileName = "suggestions"
+	tempDirName         = "tmp"
+)
+
+func main() {
+	if len(os.Args) != 4 {
+		fmt.Printf("Must provide 2 arguments: \n\t- Plugin go.mod file path\n\t- Glooe dependencies file path\n\t- merge attempts plugin go.mod file\n")
+		os.Exit(1)
+	}
+
+	pluginsModuleFilePath := os.Args[1]
+	glooDependenciesFilePath := os.Args[2]
+	var (
+		mergeAttempt    int
+		nonMatchingDeps []checks.DependencyInfoPair
+		err             error
+	)
+	if mergeAttempt, err = strconv.Atoi(os.Args[3]); err != nil {
+		fmt.Printf("Provided 3th arguments is not a number\n")
+		os.Exit(1)
+	}
+
+	if nonMatchingDeps, err = resolveDependencies(pluginsModuleFilePath, glooDependenciesFilePath, mergeAttempt); err != nil {
+		fmt.Printf("Failed to resolve dependencies: %s/n", err.Error())
+		os.Exit(1)
+	}
+
+	if len(nonMatchingDeps) == 0 {
+		fmt.Println("All shared dependencies match")
+		os.Exit(0)
+	}
+	fmt.Printf("Plugin and Gloo Enterprise dependencies do not match after %d merge attempts!\n", mergeAttempt)
+
+	// 1. Write the report to stdout
+	reportBytes, err := json.MarshalIndent(nonMatchingDeps, "", "  ")
+	if err != nil {
+		fmt.Printf("Failed to marshall error report to JSON: %s/n", err.Error())
+		os.Exit(1)
+	}
+	fmt.Println(string(reportBytes))
+
+	// 2. Write the report to a file
+	fmt.Printf("Writing error report file [%s] after %d merge attempts\n", errorReportFile, mergeAttempt)
+	if err := ioutil.WriteFile(errorReportFile, reportBytes, 0644); err != nil {
+		fmt.Printf("Failed to write error report file: %s/n", err.Error())
+	}
+
+	// 3. Create a file with suggested changes to go.mod
+	if err := createSuggestionsFile(nonMatchingDeps); err != nil {
+		fmt.Printf("Failed to create suggestions file: %s/n", err.Error())
+	}
+	os.Exit(1)
+}
+
+func resolveDependencies(moduleFilePath, glooDependenciesFilePath string, mergeAttempt int) ([]checks.DependencyInfoPair, error) {
+	var (
+		nonMatchingDeps []checks.DependencyInfoPair
+		moduleInfo      *checks.ModuleInfo
+		err             error
+	)
+	suggestionModuleFileName := moduleFilePath
+	for i := 1; mergeAttempt > 0 && i <= mergeAttempt; i++ {
```

Comment:
We should be able to write this to perform a single pass through the gloo dependencies to update the `go.mod`. 

There are three types of entries in the gloo `dependencies` file:

1. Pinned dependency: 
`bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898`
2. Dependency with version replaced: 
`github.com/Azure/go-autorest v12.0.0+incompatible => github.com/Azure/go-autorest v13.0.0+incompatible`
3. Dependency with version and name replaced:
`github.com/docker/docker v1.13.1 => github.com/moby/moby v0.7.3-0.20190826074503-38ab9da00309`

The simplest approach to the resolution would be: 
* Case 1: If it already requires the module at the expected version, skip. Otherwise, write a replace statement to the updated version. 
* Case 2 and 3: Make sure the replacements for the corresponding modules match `glooe`, updating or adding them as necessary. 

For (1), we could alternatively update the `require` to pull in the new version. But it might be easier just to get this working by rewriting the replaces.

Reply:
@rickducott 
If I split the line in the gloo `depedency.txt` (by spaces), then there are 2 sized slices:
- 2 long meaning a require dependency
- 5 long meaning a replace dependency
And in both slices the first element is unique.

Then doing the same with the plugin `go.mod` for the `require` section and for the `replace` section.

It is safe to merge the 2 sized slice with the `requrie` section and merge the 5 sized slice with the `replace` section?

Reply:
I'm a bit concerned about the merge into the `require` section, because requirements that come from `solo-projects` would get stripped away, while replacements that come from `solo-projects` should be preserved. We should be able to test this.

Reply:
Ah. So What are the rules regarding merging  the `require` section?
I've updated the PR (again :-) )and added test cases. So if the rules are clear. It easy to implement and test it.